Added functionality for creating new Training Libraries

omChauhanDev opened the pull request:
## What this PR does
This pr added functionality for creating new training libraries through dashboard.

Tasks :

- [x] Implemented Validations at frontend.

- [x] Applied access rights : Only for admin.

- [x] Displaying success, custom error notifications to user.

## Screenshots
### Before:
![image](https://github.com/WikiEducationFoundation/WikiEduDashboard/assets/119471995/40430225-2bfb-4e0f-8533-7e61adf1137a)

### After:
#### 1. Button on main training page
![image](https://github.com/WikiEducationFoundation/WikiEduDashboard/assets/119471995/05bb24dc-95f6-4ce3-804b-cce7267f56b2)

#### 2. Filling up form
![image](https://github.com/WikiEducationFoundation/WikiEduDashboard/assets/119471995/fd5d8335-a4a3-4d1e-84d5-dab05feabd24)

#### 3. Successfully created library
![image](https://github.com/WikiEducationFoundation/WikiEduDashboard/assets/119471995/da31296e-7e75-4d08-9d85-78d73e2669d2)

#### 4. First Library
![image](https://github.com/WikiEducationFoundation/WikiEduDashboard/assets/119471995/5aa1ad81-364f-4144-b87e-964c06b4b537)

#### 4. Validations at front-end
![image](https://github.com/WikiEducationFoundation/WikiEduDashboard/assets/119471995/04863689-448f-4f5e-986f-044d40170ec9)

#### 5. Custom Error Message 
![image](https://github.com/WikiEducationFoundation/WikiEduDashboard/assets/119471995/716e78ab-3f7c-4269-9d8c-02aebb6a45e7)

TheTrio commented on app/assets/javascripts/actions/training_modification_actions.js:
```diff
@@ -0,0 +1,35 @@
+      body: JSON.stringify(library),
+    });
+    setSubmitting(false);
+    modalHandler(false);
```
`modalHandler` [doesn't take any arguments](https://github.com/WikiEducationFoundation/WikiEduDashboard/pull/5810/files#diff-ebc286207cbdcfed824d6b79fd1d3086737bcd6cc94880d79a31181968f20943R11-R14) — I think this is in error?

omChauhanDev replied:
yes, now i made it correct

TheTrio commented on app/assets/javascripts/training/components/modals/create_library.jsx:
```diff
@@ -0,0 +1,123 @@
+  const validateFields = () => {
+    let valid = true;
+    const message = I18n.t('training.validation_message');
+    if (library.name.trim() === '') {
```
`!library.name.trim()`

TheTrio replied:
same for the others. nit, obviously

omChauhanDev replied:
sure

TheTrio commented on app/assets/javascripts/training/components/modals/create_library.jsx:
```diff
@@ -0,0 +1,123 @@
+                placeholder={`${I18n.t('training.enter')} ${I18n.t('training.library_slug')}`}
+              />
+              <button className="button light" onClick={props.modalHandler}>{I18n.t('training.cancel')}</button>
+              <span className="validation-error"> &nbsp; {firstErrorMessage || '\xa0'}</span>
```
What's the `\xa0` for?

omChauhanDev replied:
**\xa0:** This is the non-breaking space character. It's used here to ensure that there is always at least one character's worth of space in the span, even if 'firstErrorMessage' is empty. This avoids any layout issues that might occur if the span were completely empty.
**Non-Breaking Space:** Unlike a regular space character, a non-breaking space won't cause line breaks. This makes it useful in situations where you want to ensure that spacing doesn't affect the flow of text or layout in unintended ways

TheTrio commented on app/assets/javascripts/training/components/training_library_handler.jsx:
```diff
@@ -0,0 +1,29 @@
+
+  const modalHandler = () => {
+    setShowModal(!showModal);
+  };
```
`toggleModal` would be a better name here I think

omChauhanDev replied:
sure

TheTrio commented:
I think having a UI similar to this(which doesn't lose all the info you entered if the library with the same name already exists) would be better.

![image](https://github.com/WikiEducationFoundation/WikiEduDashboard/assets/10794178/a4ad2bef-ed75-4091-a766-9dda549d49cb)

ragesoss commented:
Yes, it would be nice to follow the same patter as the course creation modal, and to validate from the form so that errors can be fixed without starting over.

omChauhanDev commented:
okay, working on this

omChauhanDev commented:
I have updated the validation pattern, kindly check this out :

https://github.com/WikiEducationFoundation/WikiEduDashboard/assets/119471995/421aad36-6537-41e9-b553-076e857b2480

TheTrio commented on app/controllers/training_library_controller.rb:
```diff
@@ -0,0 +1,20 @@
+    if training_library.save
+      render json: { status: 'success', data: training_library }, status: :created
+    else
+      render json: { status: 'error', message: training_library.errors.full_messages },
```
why's this called `message` if its an array of messages?

omChauhanDev replied:
i have renamed it to 'errorMessages'

TheTrio commented on app/assets/javascripts/actions/training_modification_actions.js:
```diff
@@ -0,0 +1,54 @@
+    } else if (lowercaseMessage.includes('introduction')) {
+      dispatch(setInvalid('introduction', message));
+    } else {
+      dispatch({ type: API_FAIL, data: error });
```
We probably should only dispatch an `API_FAIL` event only once — so I recommend switching to a `for..of` loop and doing an early return in the `else` block.

omChauhanDev replied:
okay

TheTrio replied:
yes, this works too. We actually just disable that rule whenever we want to use `for...of`

TheTrio commented on app/assets/javascripts/actions/training_modification_actions.js:
```diff
@@ -0,0 +1,54 @@
+const createLibraryPromise = async (library, setSubmitting) => {
+    const response = await request('/training_library', {
+      method: 'POST',
+      headers: {
+        'Content-Type': 'application/json',
+      },
```
This is the default, so we can remove this

omChauhanDev replied:
done

TheTrio commented on app/models/training_library.rb:
```diff
@@ -23,8 +23,9 @@ class TrainingLibrary < ApplicationRecord
   serialize :translations, Hash
 
   validates_uniqueness_of :slug, case_sensitive: false
+  validates_uniqueness_of :name, case_sensitive: false
```
why was this added?

omChauhanDev replied:
in order to make sure that two libraries won't contain same name.

![image](https://github.com/WikiEducationFoundation/WikiEduDashboard/assets/119471995/cee3f328-fc8a-4d80-bf2e-398ce2dcafb8)

ragesoss commented:
I tested this out, and it's working pretty smoothly. One improvement would be to refresh the page after creating a new Library or Category, as the new one doesn't appear until after a page refresh.

omChauhanDev commented:
Hi @ragesoss and @TheTrio,

I wanted to bring this to your attention. I have implemented 'switching between different modes' using session storage and have also modified all the relevant test cases. I'm considering pushing this commit in a separate PR to maintain a clear separation of concerns.

Please let me know if you have any feedback.

ragesoss commented:
@omChauhanDev i created this branch with the earlier commits: https://github.com/WikiEducationFoundation/WikiEduDashboard/tree/Om-Training

You should open the separate PR against that branch, for dealing with merge conflicts.
